### PR TITLE
New module to handler with SOAP logs

### PR DESCRIPTION
**Pull request by jelas99**
## Description
New module to handler with SOAP logs

## Motivation and Context
This change allows to add a soap handler to log the request/responses

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

File: logbook-jaxws/pom.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+			<groupId>org.zalando</groupId>
+			<artifactId>logbook-api</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.zalando</groupId>
+			<artifactId>logbook-core</artifactId>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>jakarta.xml.ws</groupId>
```
Comment (whiskeysierra) [inline logbook-jaxws/pom.xml:23]: Wouldn't this be better? https://search.maven.org/artifact/javax.xml.ws/jaxws-api

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            MimeHeader lObject = (MimeHeader) all.next();
+            httpHeaders = httpHeaders.update(lObject.getName(), Collections.singletonList(lObject.getValue()));
+        }
+        return httpHeaders;
+    }
+
+    @Override
+    public String getContentType() {
+        String[] contentType = context.getMessage().getMimeHeaders().getHeader(HEADER_CONTENT_TYPE);
+        return Optional.ofNullable(contentType[0]).orElse("");
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java:59]: Could the array be empty?

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        String encoding = Optional.ofNullable((String) context.get(SOAPMessage.CHARACTER_SET_ENCODING))
+                .orElse(DEFAULT_CHARSET);
+        return Charset.forName(encoding);
+    }
+
+    @Override
+    public byte[] getBody() throws IOException {
+        State st = state.get().buffer(context.getMessage());
+        UnaryOperator<State> twoDigits = (v) -> st;
+        return state.updateAndGet(twoDigits).getBody();
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java:73]:
```suggestion
        return state.updateAndGet(throwingUnaryOperator(state ->
                state.buffer(context.getMessage()))).getBody();
```

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import javax.xml.soap.MimeHeader;
+import javax.xml.soap.MimeHeaders;
+import javax.xml.soap.SOAPException;
+import javax.xml.soap.SOAPMessage;
+import javax.xml.ws.handler.soap.SOAPMessageContext;
+
+import org.zalando.logbook.HttpHeaders;
+import org.zalando.logbook.HttpMessage;
+
+public abstract class AbstractMessage implements HttpMessage {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java:21]:
```suggestion
abstract class AbstractMessage implements HttpMessage {
```

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LocalRequest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+import javax.xml.namespace.QName;
+import javax.xml.ws.BindingProvider;
+import javax.xml.ws.handler.MessageContext;
+import javax.xml.ws.handler.soap.SOAPMessageContext;
+
+import org.zalando.logbook.HttpRequest;
+import org.zalando.logbook.Origin;
+
+public final class LocalRequest extends AbstractMessage implements HttpRequest {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LocalRequest.java:15]:
```suggestion
final class LocalRequest extends AbstractMessage implements HttpRequest {
```

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LocalResponse.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import java.io.IOException;
+import java.util.Optional;
+
+import javax.xml.ws.handler.MessageContext;
+import javax.xml.ws.handler.soap.SOAPMessageContext;
+
+import org.zalando.logbook.HttpResponse;
+import org.zalando.logbook.Origin;
+
+public final class LocalResponse extends AbstractMessage implements HttpResponse {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LocalResponse.java:12]:
```suggestion
final class LocalResponse extends AbstractMessage implements HttpResponse {
```

File: logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+import org.zalando.logbook.DefaultSink;
+import org.zalando.logbook.HttpLogWriter;
+import org.zalando.logbook.Logbook;
+import org.zalando.logbook.Logbook.ResponseProcessingStage;
+import org.zalando.logbook.TestStrategy;
+
+@ExtendWith(MockitoExtension.class)
+@DisplayNameGeneration(ReplaceUnderscores.class)
+@TestMethodOrder(OrderAnnotation.class)
+public class LoggingHandlerTest {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java:47]: A local integration test with a real network roundtrip would be give me more confidence.

File: logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+import org.zalando.logbook.DefaultHttpLogFormatter;
+import org.zalando.logbook.DefaultSink;
+import org.zalando.logbook.HttpLogWriter;
+import org.zalando.logbook.Logbook;
+import org.zalando.logbook.Logbook.ResponseProcessingStage;
+import org.zalando.logbook.TestStrategy;
+
+@ExtendWith(MockitoExtension.class)
+@DisplayNameGeneration(ReplaceUnderscores.class)
+@TestMethodOrder(OrderAnnotation.class)
```
Comment (whiskeysierra) [inline logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java:46]: Why?

File: logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        lenient().when(context.get(MessageContext.WSDL_OPERATION)).thenReturn(new QName("operation"));
+        lenient().when(context.get(BindingProvider.ENDPOINT_ADDRESS_PROPERTY))
+                .thenReturn("http://localhost:8080/soap/operation");
+        lenient().when(context.get(SOAPMessage.CHARACTER_SET_ENCODING)).thenReturn("UTF-8");
+        lenient().when(context.get(SOAPMessageContext.QUERY_STRING)).thenReturn("id=12345");
+
+        // -- underTest
+        boolean result = underTest.handleMessage(context);
+
+        // -- assertions
```
Comment (whiskeysierra) [inline logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java:74]: Those assertions are not really testing anything important, are they?

File: logbook-jaxws/pom.xml (hunk 1 above)
Comment (jelas99) [reply logbook-jaxws/pom.xml:23]: I think not. Oracle decided to transfer Java EE fully to an open source foundation (Eclipse). If you open the home page in the link that you shared you can see that information

Comment (whiskeysierra) [reply logbook-jaxws/pom.xml:23]: Ah, good to know!

File: logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java (hunk 7 above)
Comment (jelas99) [reply logbook-jaxws/src/test/java/org/zalando/logbook/jaxws/LoggingHandlerTest.java:47]: Good point. I can handle with that. 👍

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+    protected final AtomicReference<State> state = new AtomicReference<>(new Unbuffered());
+    protected SOAPMessageContext context;
+
+    protected AbstractMessage(SOAPMessageContext context) {
+        this.context = context;
+    }
+
+    @Override
+    public String getProtocolVersion() {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/AbstractMessage.java:34]: This should return either `"HTTP/1.0", "HTTP/1.1" or "HTTP/2".

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LoggingHandler.java [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+
+import javax.xml.namespace.QName;
+import javax.xml.ws.handler.MessageContext;
+import javax.xml.ws.handler.soap.SOAPHandler;
+import javax.xml.ws.handler.soap.SOAPMessageContext;
+
+import org.zalando.logbook.Logbook;
+import org.zalando.logbook.Logbook.ResponseProcessingStage;
+
+public class LoggingHandler implements SOAPHandler<SOAPMessageContext> {
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LoggingHandler.java:14]:
```suggestion
public final class LoggingHandler implements SOAPHandler<SOAPMessageContext> {
```

File: logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LoggingHandler.java [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+    @Override
+    public boolean handleMessage(SOAPMessageContext context) {
+        boolean outbound = (boolean) context.get(MessageContext.MESSAGE_OUTBOUND_PROPERTY);
+
+        if (outbound) {
+            LocalRequest request = new LocalRequest(context);
+            try {
+                ResponseProcessingStage process = logbook.process(request).write();
+                STAGE_CONTEXT.set(process);
```
Comment (whiskeysierra) [inline logbook-jaxws/src/main/java/org/zalando/logbook/jaxws/LoggingHandler.java:31]: Can't you put the stage into the context? That would remove the need to have a ThreadLocal here.

**Comment by whiskeysierra** [comment]
@jelas99 I took your branch as a blue print and started to work on top. 
I rewrote most of it and started with an integration-style test.

I'll close this PR in favor of https://github.com/zalando/logbook/tree/feature/jaxws.

Once I'm done with tests, I'll create a new pull request.

**Pull request closed without merging**